apps/prairielearn/elements/pl-file-upload/pl-file-upload.py

```diff
@@ -21,28 +28,133 @@ def get_file_names_as_array(raw_file_names: str) -> list[str]:
     return next(reader)
 
 
-# Each pl-file-upload element is uniquely identified by the SHA1 hash of its
-# file_names attribute
-def get_answer_name(file_names: str) -> str:
-    return "_file_upload_{0}".format(
-        hashlib.sha1(file_names.encode("utf-8")).hexdigest()
+def match_regex_with_files(
+    regex_patterns: list[str], files_names: list[str], limit_1: bool
+) -> tuple[list[str], list[str]]:
+    """
+    Takes a list of regexes (as returned by extract_patterns) and a list of file names and matches them 1:n or 1:1, depending on limit_1 parameter
+    Returns a tuple of matched file names and unmatched patterns
+    """
+    unmatched_patterns = regex_patterns.copy()
+    remaining_files = files_names.copy()
```

Comment:
Unfortunately, I cannot use sets for either the files nor the patterns here.

The order in which I iterate over `remaining-files` can affect the output of the function as it determines which patterns get matched with which file. Cases where patterns overlap and this order matters should be avoided by instructors (and I'll make sure to add a warning in the documentation), but I think that the code should nonetheless produce somewhat coherent results for those cases rather than ones that are entirely determined by hash sorting.

For `unmatched_patterns`, the order does not matter, but I do need to allow duplicates, as repeated patterns like `file-patterns="*.py,*.py"` would be an actually acceptable way to define "exactly 2 files ending in .py".